spock-core/src/main/java/org/spockframework/runtime/ExtensionRunner.java

```diff
@@ -163,7 +160,12 @@ private IAnnotationDrivenExtension getAnnotationDrivenExtension(Annotation annot
   }
 
 
+  @SuppressWarnings("unchecked")
   private IAnnotationDrivenExtension getOrCreateExtension(Class<? extends IAnnotationDrivenExtension> clazz) {
+    if (IStatelessAnnotationDrivenExtension.class.isAssignableFrom(clazz)) {
+      // we need to add the extension to the local extensions map to ensure that `visitSpec` is called for it at the end
+      return localExtensions.computeIfAbsent(clazz, __ -> extensionRegistry.getStatelessAnnotationDrivenExtension((Class<? extends IStatelessAnnotationDrivenExtension<?>>) clazz));
```

Comment:
```suggestion
      return localExtensions.computeIfAbsent(clazz, c -> extensionRegistry.getStatelessAnnotationDrivenExtension((Class<? extends IStatelessAnnotationDrivenExtension<?>>) c));
```

Allows that the JVM does not need to instaniate the lambda for each call separatly, because it does not capture an outer argument.